chainer/functions/array/tile.py

```diff
@@ -0,0 +1,75 @@
+import itertools
+import six
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _tile_indices(reps, shape):
+    """Return tile indices from reps and shape of input array."""
+    tile_indices = []
+    ranges = [six.moves.range(x) for x in reps]
+    for index in itertools.product(*ranges):
+        tile_index = tuple(slice(i * d, (i + 1) * d)
+                           for i, d in zip(index, shape))
+        tile_indices.append(tile_index)
+    return tile_indices
+
+
+class Tile(function.Function):
+    """Tiling of an array."""
+
+    def __init__(self, reps):
+        if isinstance(reps, six.integer_types):
+            self.reps = (reps,)
+        elif isinstance(reps, tuple) and all(
+                isinstance(x, six.integer_types) for x in reps):
+            self.reps = reps
+        else:
+            raise TypeError('reps must be int or tuple of ints')
+
+        if not all(x >= 0 for x in self.reps):
+            raise ValueError('all elements in reps must be zero or larger')
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+
+    def forward(self, inputs):
+        xp = cuda.get_array_module(*inputs)
+        return xp.tile(inputs[0], self.reps),
+
+    def backward(self, inputs, grads):
+        xp = cuda.get_array_module(*inputs)
+        x = inputs[0]
+        reps = self.reps
+
+        # Ensure input and reps have the same length.
+        if x.ndim > len(reps):
+            reps = (1,) * (x.ndim - len(reps)) + reps
+        elif x.ndim < len(reps):
+            x = x.reshape((1,) * (len(reps) - x.ndim) + x.shape)
+
+        gy = xp.zeros(x.shape).astype(x.dtype, copy=False)
+        for index in _tile_indices(reps, x.shape):
+            gy += grads[0][index]
```

Comment:
I checked the difference of performance between for-loop and sum versions, the latter being always faster:

```
xp:<module 'numpy' from '/home/fujita/.pyenv/versions/anaconda3-4.0.0/lib/python3.5/site-packages/numpy/__init__.py'> rep0:2 rep1:2 size:2
tile_loop: 0.00018930435180664062
tile_sum: 0.00016498565673828125
xp:<module 'numpy' from '/home/fujita/.pyenv/versions/anaconda3-4.0.0/lib/python3.5/site-packages/numpy/__init__.py'> rep0:2 rep1:2 size:1000
tile_loop: 0.0001862049102783203
tile_sum: 0.00016951560974121094
xp:<module 'numpy' from '/home/fujita/.pyenv/versions/anaconda3-4.0.0/lib/python3.5/site-packages/numpy/__init__.py'> rep0:2 rep1:100 size:2
tile_loop: 0.0006461143493652344
tile_sum: 0.00016927719116210938
xp:<module 'numpy' from '/home/fujita/.pyenv/versions/anaconda3-4.0.0/lib/python3.5/site-packages/numpy/__init__.py'> rep0:2 rep1:100 size:1000
tile_loop: 0.001209259033203125
tile_sum: 0.0006072521209716797
xp:<module 'numpy' from '/home/fujita/.pyenv/versions/anaconda3-4.0.0/lib/python3.5/site-packages/numpy/__init__.py'> rep0:100 rep1:2 size:2
tile_loop: 0.000652313232421875
tile_sum: 0.00016951560974121094
xp:<module 'numpy' from '/home/fujita/.pyenv/versions/anaconda3-4.0.0/lib/python3.5/site-packages/numpy/__init__.py'> rep0:100 rep1:2 size:1000
tile_loop: 0.0015964508056640625
tile_sum: 0.0006909370422363281
xp:<module 'numpy' from '/home/fujita/.pyenv/versions/anaconda3-4.0.0/lib/python3.5/site-packages/numpy/__init__.py'> rep0:100 rep1:100 size:2
tile_loop: 0.02381134033203125
tile_sum: 0.00037789344787597656
xp:<module 'numpy' from '/home/fujita/.pyenv/versions/anaconda3-4.0.0/lib/python3.5/site-packages/numpy/__init__.py'> rep0:100 rep1:100 size:1000
tile_loop: 0.048241376876831055
tile_sum: 0.014247894287109375
xp:<module 'cupy' from '/home/fujita/chainer/cupy/__init__.py'> rep0:2 rep1:2 size:2
tile_loop: 0.0004968643188476562
tile_sum: 0.00044727325439453125
xp:<module 'cupy' from '/home/fujita/chainer/cupy/__init__.py'> rep0:2 rep1:2 size:1000
tile_loop: 0.00047969818115234375
tile_sum: 0.0003783702850341797
xp:<module 'cupy' from '/home/fujita/chainer/cupy/__init__.py'> rep0:2 rep1:100 size:2
tile_loop: 0.004094123840332031
tile_sum: 0.00037384033203125
xp:<module 'cupy' from '/home/fujita/chainer/cupy/__init__.py'> rep0:2 rep1:100 size:1000
tile_loop: 0.004076242446899414
tile_sum: 0.0003809928894042969
xp:<module 'cupy' from '/home/fujita/chainer/cupy/__init__.py'> rep0:100 rep1:2 size:2
tile_loop: 0.00402069091796875
tile_sum: 0.0003752708435058594
xp:<module 'cupy' from '/home/fujita/chainer/cupy/__init__.py'> rep0:100 rep1:2 size:1000
tile_loop: 0.004088878631591797
tile_sum: 0.0003764629364013672
xp:<module 'cupy' from '/home/fujita/chainer/cupy/__init__.py'> rep0:100 rep1:100 size:2
tile_loop: 0.18110251426696777
tile_sum: 0.0003802776336669922
xp:<module 'cupy' from '/home/fujita/chainer/cupy/__init__.py'> rep0:100 rep1:100 size:1000
tile_loop: 0.18401122093200684
tile_sum: 0.00038433074951171875
```
Here is the code I used for this evaluation:
```
import time
import chainer.functions.array.tile_loop as tile_loop
import chainer.functions.array.tile_sum as tile_sum
from chainer import functions as F
import cupy
import numpy as np


for xp in (np, cupy):
    for rep0 in (2, 100):
        for rep1 in (2, 100):
            for size in (2, 1000):
                x = xp.arange(size, dtype=np.float32)
                print('xp:{} rep0:{} rep1:{} size:{}'.format(xp, rep0, rep1, size))

                reps = (rep0, rep1)

                F.sum(tile_loop.tile(x, reps)).backward()
                t = time.time()
                F.sum(tile_loop.tile(x, reps)).backward()
                print('tile_loop:', time.time() - t)

                F.sum(tile_sum.tile(x, reps)).backward()
                t = time.time()
                F.sum(tile_sum.tile(x, reps)).backward()
                print('tile_sum:', time.time() - t)
```